sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -577,23 +577,31 @@ public void setWindowedWrites(boolean windowedWrites) {
      * idempotent, as it may be executed multiple times in the case of failure or for redundancy. It
      * is a best practice to attempt to try to make this method atomic.
      *
+     * <p>Returns the set of temporary files generated. Callers must call
+     * {@link #removeTemporaryFiles(Set)} to cleanup these files.
+     *
      * @param writerResults the results of writes (FileResult).
      */
-    public void finalize(Iterable<FileResult<DestinationT>> writerResults) throws Exception {
+     public Set<ResourceId> finalize(
+         Iterable<FileResult<DestinationT>> writerResults) throws Exception {
       // Collect names of temporary files and rename them.
       Map<ResourceId, ResourceId> outputFilenames = buildOutputFilenames(writerResults);
       copyToOutputFiles(outputFilenames);
+       return outputFilenames.keySet();
+    }
 
-      // Optionally remove temporary files.
-      // We remove the entire temporary directory, rather than specifically removing the files
-      // from writerResults, because writerResults includes only successfully completed bundles,
-      // and we'd like to clean up the failed ones too.
-      // Note that due to GCS eventual consistency, matching files in the temp directory is also
-      // currently non-perfect and may fail to delete some files.
-      //
-      // When windows or triggers are specified, files are generated incrementally so deleting
-      // the entire directory in finalize is incorrect.
-      removeTemporaryFiles(outputFilenames.keySet(), !windowedWrites);
+    /* Remove temporary files after finalization.
+     *
+     * <p>We remove the entire temporary directory, rather than specifically removing the files from
+     * writerResults, because writerResults includes only successfully completed bundles, and we'd
+     * like to clean up the failed ones too. Note that due to GCS eventual consistency, matching
+     * files in the temp directory is also currently non-perfect and may fail to delete some files.
+     *
+     * <p>When windows or triggers are specified, files are generated incrementally so deleting the
```

Comment:
Which part of the comment is incorrect? Note that we never use the optimization of deleting the entire directory if windowedWrites==true because many parallel triggers, windows, etc. might be writing files at the same time to the same directory.